cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStore.java

```diff
@@ -0,0 +1,544 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * File based implementation of secure store. Uses Java JCEKS based keystore.
+ * The data and its metadata are both stored in the keystore separately.
+ * The data is stored with the provided name as the key and the metadata is
+ * stored with the provided name + "_metadata" as the key. They are stored separately
+ * because they are accessed separately most of the time.
+ *
+ * When the client calls a put, the key is put in the keystore, the metadata object
+ * for that key is created and that is put in the store too. The system then flushes
+ *  the keystore to the file system.
+ *  During the flush, the current file is first backed up (_OLD). Then the keystore
+ *  is written to temporary file (_NEW). If that is successful then the temporary
+ *  file is renamed to the secure store file. If anything fails during this process
+ *  then the keystore reverts to the last  successfully written file.
+ *
+ *  The keystore is flushed to the filesystem after every put and delete.
+ */
+@Singleton
+class FileSecureStore implements SecureStore, SecureStoreManager {
+  private static final Logger LOG = LoggerFactory.getLogger(FileSecureStore.class);
+
+  private static final String SCHEME_NAME = "jceks";
+  /*
+   This is used to create the key for storing metadata in the keystore. Changing this will make the
+   metadata for existing entries unreachable.
+   */
+  private static final String METADATA_SUFFIX = "_metadata";
+  private static final String DEFAULT_FILE_PATH = "/tmp";
+  private static final String DEFAULT_FILE_NAME = "securestore";
+  private static final char[] DEFAULT_PASSWORD = "cdapsecret".toCharArray();
+  /*
+   Java Keystore needs an algorithm name to store a key, it is not used for any checks, only stored,
+   since we are not handling the encryption we don't care about this.
+  */
+  private static final String ALGORITHM_PROXY = "none";
+
+  private final char[] password;
+  private final Path path;
+  private final Lock readLock;
+  private final Lock writeLock;
+
+  private KeyStore keyStore;
```

Comment:
final. You can do so by having the `locateKeystore` method returns the `KeyStore`.